### PR TITLE
Completely republish QoS 2 messages when reconnecting clean-session clients

### PR DESCRIPTION
**Pull request by amrbekhit**
Modified _messages_reconnect_reset_out() to completely republish QoS 2 messages when a clean-session client reconnects.

This eliminates issue #276 whereby clean-session clients would previously attempt to resume transmission of QoS 2 messages where they left off. However, because the server does not save state for such clients, the messages are unexpected and causes the server to close the connection (observed using mosquitto-1.4.15). This then continuously prevents the client from reconnecting.

loop_misc() has also been modified to only call _message_retry_check() if the client is connected. This prevents the client from immediately retrying transmission after sending a CONNECT message but before receiving CONNACK from the server. The out-of-order messages again causes the server to close the connection, preventing the client from reconnecting.

### CODE REVIEW

**Comment by vrst37** [comment]
You will need to sign Eclipse contributor agreement (https://www.eclipse.org/legal/ECA.php) and have a signed-off commit (use commit -s) before this commit can be accepted. 

Also, these changes have failed the test on Travis. I think these would have to be fixed before the changes can be merged.

**Comment by amrbekhit** [comment]
I've made a new commit that modifies the tests to reflect the changes made in my original commit. Essentially, the original code used to handle QoS2 reconnects the same, regardless of whether the client is using a clean session or not, and the original tests were written based on that. My patch corrects this issue so that QoS2 reconnects are now different depending on the clean session state. I've modified the existing test and added a new one to test both of these situations.

**Comment by PierreF** [comment]
I disagree with the with on loop_misc fix. The client is perfectly allowed to immediately start sending packet. From MQTT 3.1.1 spec:

> Clients are allowed to send further Control Packets immediately after sending a CONNECT Packet; Clients need not wait for a CONNACK Packet to arrive from the Server.

I'm not sure to agree on the way to fix this issue for QoS 2 full resend. I fear that in some case it could result in two delivery of the message (publish, pubrec, pubrel, server send the pubcomp but connection break before puback reach the client).
In addition to my understanding this is NOT compliant with MQTT specification which state:

> If CleanSession is set to 1, the Client and Server MUST discard any previous Session
> [...]
> The Session state in the Client consists of:
> * QoS 1 and QoS 2 messages which have been sent to the Server, but have not been completely acknowledged.

So from the standard, we must discard any QoS 1&2 message for which the publish were sent to the server in case of reconnection with clean session.
I agree that we are probably not compliant for QoS 1 packet and resend them. But going with a non-standard behavior should not cause unexpected result (QoS 2, Exactly once delivery, but could cause multiple delivery). And yes the standard means that with cleanSession=1 you may get zero delivery.

To summarize:
* the fix on loop_misc must not be needed. It's a broker bug.
* We should follow standard cleanSession=1 handling (probably with some documentation reminding that cleansession=1 means possible lost)
* If going with non-standard behavior, it should be optional, documented as such and should avoid possible multiple delivery of QoS message 2.
